src/core/QCheck_ng.ml

```diff
@@ -0,0 +1,1966 @@
+(*
+QCheck: Random testing for OCaml
+copyright (c) 2013-2017, Guillaume Bury, Simon Cruanes, Vincent Hugot, Jan Midtgaard
+all rights reserved.
+*)
+
+(** {1 Quickcheck inspired property-based testing} *)
+
+let poly_compare=compare
+open Printf
+
+module RS = Random.State
+
+let (|>) x f = f x
+
+let rec foldn ~f ~init:acc i =
+  if i = 0 then acc else foldn ~f ~init:(f acc i) (i-1)
+
+let _is_some = function Some _ -> true | None -> false
+
+let _opt_map_2 ~f a b = match a, b with
+  | Some x, Some y -> Some (f x y)
+  | _ -> None
+
+let _opt_map_3 ~f a b c = match a, b, c with
+  | Some x, Some y, Some z -> Some (f x y z)
+  | _ -> None
+
+let _opt_map_4 ~f a b c d = match a, b, c, d with
+  | Some x, Some y, Some z, Some w -> Some (f x y z w)
+  | _ -> None
+
+let _opt_sum a b = match a, b with
+  | Some _, _ -> a
+  | None, _ -> b
+
+let sum_int = List.fold_left (+) 0
+
+exception FailedPrecondition
+(* raised if precondition is false *)
+
+exception No_example_found of string
+(* raised if an example failed to be found *)
+
+let assume b = if not b then raise FailedPrecondition
+
+let assume_fail () = raise FailedPrecondition
+
+let (==>) b1 b2 = if b1 then b2 else raise FailedPrecondition
+
+module Seq = struct
+
+  include Seq
```

Comment:
Are you sure that you want include `Seq` from `Stdlib` in this... "internal Seq`?

Reply:
No, I am not sure :D The idea was to extend the Stdlib `Seq` and add a few utilites that may or may not be there (like `unfold` or `cons`) without having to care about it in the rest of the code.

This may or may not be a bad idea, I don't have enough experience in OCaml to have an opinion.

Your question suggest this is a bad idea, I don't mind changing that, but for my culture, what pros/cons do you see to this trick please?

Reply:
But do you want "extend Seq" in the public API or just "internally" ?
My guess is, like it is done for opt, you can have internal functions (in a module or not) and for example, if you d'ont want export-it without writing an mli ;) you can use this kind of trick : 

```ocaml
open struct 
    ...
end
```
But since you can have a `Seq` module and refer the older one, you can use `Stdlib.Seq` or maintaining an alias of Seq.

Reply:
Oh, now I understand your question :smile: This is definitely an internal module, this will never appear in the MLI, there's no need/reason to export Seq or any such function I think

Reply:
Without a `.mli` the intention is unclear... ;)

Reply:
You can include `Seq` in the .ml, but writing a fixed API in the .mli (if the module is exposed at all) is not as good on a stability standpoint. In the .ml I don't see objections :-)

Reply:
Now that I have added the MLI, is it clearer? Do you still think there should be some changes?

Reply:
Yes. But I don't know how to resolve the comment.